src/game/lara/lara_cheat.c

```diff
@@ -352,34 +352,38 @@ bool Lara_Cheat_Teleport(int32_t x, int32_t y, int32_t z)
         g_Lara.gun_status = LGS_ARMLESS;
     }
 
-    const bool has_flare = Lara_Cheat_HasFlare();
     Lara_GetOffVehicle();
 
-    const ROOM_INFO *const room = &g_Rooms[g_LaraItem->room_num];
-    const bool room_submerged = (room->flags & RF_UNDERWATER) != 0;
-    const int16_t water_height = Room_GetWaterHeight(
-        g_LaraItem->pos.x, g_LaraItem->pos.y, g_LaraItem->pos.z,
-        g_LaraItem->room_num);
-
-    if (room_submerged || (water_height != NO_HEIGHT && water_height > 0)) {
-        g_Lara.water_status = LWS_UNDERWATER;
-        g_LaraItem->current_anim_state = LS_SWIM;
-        g_LaraItem->goal_anim_state = LS_SWIM;
-        g_LaraItem->anim_num =
-            g_Objects[O_LARA].anim_idx + LA_UNDERWATER_SWIM_FORWARD_DRIFT;
-        g_LaraItem->frame_num = g_Anims[g_LaraItem->anim_num].frame_base;
-    } else {
-        g_Lara.water_status = LWS_ABOVE_WATER;
-        g_LaraItem->current_anim_state = LS_STOP;
-        g_LaraItem->goal_anim_state = LS_STOP;
-        g_LaraItem->anim_num = g_Objects[O_LARA].anim_idx + LA_STAND_STILL;
-        g_LaraItem->frame_num = g_Anims[g_LaraItem->anim_num].frame_base;
-        g_LaraItem->rot.x = 0;
-        g_LaraItem->rot.z = 0;
-        g_Lara.head_x_rot = 0;
-        g_Lara.head_y_rot = 0;
-        g_Lara.torso_x_rot = 0;
-        g_Lara.torso_y_rot = 0;
+    if (g_Lara.extra_anim) {
+        const ROOM_INFO *const room = &g_Rooms[g_LaraItem->room_num];
+        const bool room_submerged = (room->flags & RF_UNDERWATER) != 0;
+        const int16_t water_height = Room_GetWaterHeight(
+            g_LaraItem->pos.x, g_LaraItem->pos.y, g_LaraItem->pos.z,
+            g_LaraItem->room_num);
+
+        if (room_submerged || (water_height != NO_HEIGHT && water_height > 0)) {
+            g_Lara.water_status = LWS_UNDERWATER;
+            g_LaraItem->current_anim_state = LS_SWIM;
+            g_LaraItem->goal_anim_state = LS_SWIM;
+            g_LaraItem->anim_num =
+                g_Objects[O_LARA].anim_idx + LA_UNDERWATER_SWIM_FORWARD_DRIFT;
+            g_LaraItem->frame_num = g_Anims[g_LaraItem->anim_num].frame_base;
+        } else {
+            g_Lara.water_status = LWS_ABOVE_WATER;
+            g_LaraItem->current_anim_state = LS_STOP;
+            g_LaraItem->goal_anim_state = LS_STOP;
+            g_LaraItem->anim_num = g_Objects[O_LARA].anim_idx + LA_STAND_STILL;
+            g_LaraItem->frame_num = g_Anims[g_LaraItem->anim_num].frame_base;
+            g_LaraItem->rot.x = 0;
+            g_LaraItem->rot.z = 0;
+            g_Lara.head_x_rot = 0;
+            g_Lara.head_y_rot = 0;
+            g_Lara.torso_x_rot = 0;
+            g_Lara.torso_y_rot = 0;
+        }
+        g_Lara.extra_anim = 0;
+        const bool has_flare = Lara_Cheat_HasFlare();
+        Lara_Cheat_ReinitialiseGunMeshes(has_flare);
```

Comment:
Thanks. I wonder if there's any way to simplify this logic?

Reply:
I've tidied it up, checking the state multiple times didn't make sense, as it's only needed just before `Lara_InitialiseMeshes` is called, as that's what resets it. Once we know more about flares we can hopefully eliminate it.